Make ClientData thread-safe

Yadunund opened the pull request:
Make access to `ClientData` thread-safe.

~~I've not fully tested this PR yet but I know `colcon test --event-handlers console_direct+ --packages-select rclcpp_action --ctest-args -R test_server` fails. It has to do with how i'm handling cases where the client is shutdown when there are still queries in flight.~~

Yadunund committed:
Make ClientData thread-safe

Signed-off-by: Yadunund <yadunund@intrinsic.ai>

ahcorde reviewed (changes requested):
fix conflicts

Yadunund committed:
Handle queries in flight

Signed-off-by: Yadunund <yadunund@intrinsic.ai>

Yadunund commented:
@ahcorde could you help check if the expected tests are passing here?

ahcorde commented:
[ros2.repos](https://gist.github.com/ahcorde/6f970ad7270f43c85d43c3444adf27f9)

# rmw_zenoh/yadu/raii-service
* rmw_zenoh commit: https://github.com/ros2/rmw_zenoh/commit/aaa01c187ab90c01d9dd55b89818f731846c1dc1

| ROS Package      | Failing test name | Failing test output     |
| :---:        |    :----:   |          :---: |
| rcl | None | None |
| rcl_action | None | None |
| rcl_lifecycle   | None        | None      |
| rcl_yaml_param_parser   | None        | None      |
| rclcpp   | test_intra_process_manager        | Failed (Check logs)      |
| rclcpp   | test_node_interfaces__node_graph        | Failed (Check logs)      |
| rclcpp   | test_publisher        | Failed (Check logs)      |
| rclcpp   | test_events_executor        | Timeout (Check logs)      |
| rclcpp_action   | None       | None      |
| rclcpp_components   | None        | None      |
| rclcpp_lifecycle  | None | -  |
| rmw_zenoh_cpp   | None        | None      |
| test_cli   | None        | None      |
| test_cli_remap   | test_cli_remapping        | TBD |
| rclpy   | None | |

clalancette commented:
Because of #298, this now needs a rebase and conflicts resolved.  Once that is done, I'll do another review pass over it.

Yadunund commented:
@clalancette i've resolved the conflicts!

clalancette commented on rmw_zenoh_cpp/src/detail/graph_cache.cpp:
```diff
@@ -1182,15 +1181,15 @@ rmw_ret_t GraphCache::get_entities_info_by_topic(
 rmw_ret_t GraphCache::service_server_is_available(
-  const char * service_name,
-  const char * service_type,
+  const liveliness::TopicInfo & client_topic_info,
```
Technically, we should `#include "liveliness_utils.hpp"` for this class.

clalancette commented on rmw_zenoh_cpp/src/rmw_zenoh.cpp:
```diff
@@ -1405,245 +1411,69 @@ rmw_create_client(
-
-  if (rmw_validate_full_topic_name(service_name, &validation_result, nullptr) != RMW_RET_OK) {
-    RMW_SET_ERROR_MSG("rmw_validate_full_topic_name failed");
+  // Get the RMW the type support.
```
```suggestion
  // Get the service type support.
```

clalancette commented on rmw_zenoh_cpp/src/rmw_zenoh.cpp:
```diff
@@ -1405,245 +1411,69 @@ rmw_create_client(
+    // Error already handled.
     return nullptr;
   }
 
```
If a step below fails, do we need to remove this client from the node?  In other words, do we need a `rcpputils::scope_exit` block here?

Yadunund replied:
Ah nope! The shared_ptr we created within `create_client_data` will go out of scope and the destructor will clean everything up. The `free_rmw_client` block will take care of deallocating the `rmw_client_t` object.

clalancette commented on rmw_zenoh_cpp/src/rmw_zenoh.cpp:
```diff
@@ -1405,245 +1411,69 @@ rmw_create_client(
   RMW_CHECK_FOR_NULL_WITH_MSG(
     rmw_client->service_name,
-    "failed to allocate client name",
+    "failed to allocate service name",
```
```suggestion
    "failed to allocate service client name",
```

clalancette commented on rmw_zenoh_cpp/src/rmw_zenoh.cpp:
```diff
@@ -2866,29 +2522,16 @@ rmw_service_server_is_available(
   }
 
-  std::string service_type = client_data->request_type_support->get_name();
-  size_t suffix_substring_position = service_type.find("Request_");
```
I notice that we are no longer checking whether the service_type has `Request_` in it.  Do we need to add that back in?

Yadunund replied:
So this check was moved inside `ClientData::make()` but there was typo there anyways which I have fixed https://github.com/ros2/rmw_zenoh/pull/293/commits/4c9b1069629bfd2daea1f3e14ef5376dfca3f356. It would not have caused any problems since the substring we extract will be the same but now it's semantically correct.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -0,0 +1,572 @@
+// This is needed as the decrement_queries_in_flight callback that me be executed
+// after ClientData is deallocated will try to lock an invalid mutex and thus lead
+// to UB. This only happens in rclcpp_action/test_client.
+static std::mutex num_in_flight_mutex;
```
I'm not sure I understand how this fixes the problem, and I'm not sure it is required anymore.  If `client_data` has gone out of scope, then calling `client_data->decrement_queries_in_flight()` is already UB, never mind accessing the `std::mutex` within it (it will usually happen to work, but that's besides the point).

But we really shouldn't get into this situation anymore anyway, because `NodeData::delete_client_data` will only be deleting the structure if there is nothing in flight.  (The potential memory leak is worse here, in that `client_data_drop` isn't cleaning up anymore if this is the "last" thing in flight, but that is not UB).

In other words, I don't see why we can't embed this mutex in the `ClientData` structure.  But there may be something I'm missing.

Yadunund replied:
I don't think I implemented this correctly.

My approach here was to avoid UBs only since memory leaks should never happen given the use of shared_ptrs.

If `client_data` is deallocated by resetting an `rclcpp::Node` ptr (as done in one of the tests), dereferencing `client_data` within `client_data_drop` like `client_data-> decrement_queries_in_flight()` would be UB. In this scenario, even though we do not remove the `client_data` from `NodeData::clients_` when `NodeData::delete_client_data()` is called, it will get removed when the `NodeData`  is deleted by the reset call. 

Ideally we have a static map of ClientData rawptr address to num_in_flight. Then `client_data_drop` would lookup this map and decrement the count. We would increment the count in the same map.

But I don't know if it worth the effort. Please let me know what you think.

Yadunund replied:
Alright I think I have a better solution implemented in https://github.com/ros2/rmw_zenoh/pull/293/commits/37e2e0372cec18665ef2fe1307a95646735711bc

Yadunund committed:
Address some feedback

Signed-off-by: Yadunund <yadunund@intrinsic.ai>

Yadunund committed:
Fix service client type check

Signed-off-by: Yadunund <yadunund@intrinsic.ai>

Yadunund committed:
fix tracking of num in flight without UB

Signed-off-by: Yadunund <yadunund@intrinsic.ai>

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -0,0 +1,613 @@
+  }
+  std::lock_guard<std::mutex> lock(num_in_flight_mutex);
+  num_in_flight_map.erase(this);
+  deleted_clients.insert(this);
```
Importantly, I think we only need to add this to the `deleted_clients` map if there is anything in flight.  I think that should keep us from "leaking" memory in the `deleted_clients` map for every deleted client.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -0,0 +1,613 @@
+  // the clients_ maps would get erased when the NodeData's destructor is invoked.
+  // This is an edge case that should be resolved once we switch to zenoh-cpp and capture
+  // weaK_ptr<ClientData> in this callback instead.
+  auto num_in_flight_it = num_in_flight_map.find(client_data);
```
This needs to take the `num_in_flight_mutex`.

clalancette commented on rmw_zenoh_cpp/src/detail/rmw_client_data.cpp:
```diff
@@ -0,0 +1,613 @@
+  free_token.cancel();
+
+  // Erase from deleted_clients set if the memory address for the client data is reused.
+  num_in_flight_map.erase(client_data.get());
+  deleted_clients.erase(client_data.get());
```
There are a few different problems with this that I see.

First, this needs to take the `num_in_flight_mutex` for mutual exclusion.

Second, I don't think that we ever need to erase from the `num_in_flight_map`.  That only exists for the lifetime of a `ClientData` object.  In point of fact, we may actually want to add this object to the map right here (with a `num_in_flight` of 0).

Finally, I don't think that erasing this from the `deleted_clients` is the right thing to do.  If we take my other suggestion on only adding to the `deleted_clients` map if queries are in-flight, then removing this here means that those outstanding queries may do the wrong thing during `client_data_handler`.  That is, they will see that they are *not* in the `deleted_clients` map, and go on to operate on a different object.  This is actually tricky to fix, though I have a somewhat hacky idea.  That idea is to continually allocate `ClientData` structures (as in line 236) until we find an address that is *not* in `deleted_clients`.  Once we find one, we deallocate all of the other ones.  That way we'll be guaranteed to get an address that does not have queries currently in flight.

Yadunund replied:
> That idea is to continually allocate ClientData structures (as in line 236) until we find an address that is not in deleted_clients. Once we find one, we deallocate all of the other ones. That way we'll be guaranteed to get an address that does not have queries currently in flight.

That should resolve it but wondering we need to go through this hassle now especially since we will undo all these changes once switch to zenoh-cpp and will rely on lambda captures?

clalancette replied:
Personally, I think it is worth it.  It's not a whole lot of code, and we never know what is going to happen in the future.  The good news is that it is not too hard; see #304, which targets this PR and fixes all of these issues (I think).

clalancette reviewed (changes requested):
Overall, I like the new direction here.  I have some specific concerns on some of the data handling inline.

clalancette committed:
Additional fixes for yadu/raii-client. (#304)

* Additional fixes for yadu/raii-client.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

* Apply suggestions from code review

Co-authored-by: yadunund <yadunund@intrinsic.ai>
Signed-off-by: Chris Lalancette <clalancette@gmail.com>

* Fixups.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

---------

Signed-off-by: Chris Lalancette <clalancette@gmail.com>
Co-authored-by: yadunund <yadunund@intrinsic.ai>

Yadunund commented:
nav2 demo crashes when the following steps are performed
1. ros2 launch nav2_bringup tb3_simulation_launch.py headless:=False
2. Localize the robot and wait for startup to complete
3. Click `Reset` from the nav2 panel in RViz

```
[component_container_isolated-8] malloc(): unsorted double linked list corrupted
[component_container_isolated-8] Magick: abort due to signal 6 (SIGABRT) "Abort"...
[ERROR] [component_container_isolated-8]: process has died [pid 2257204, exit code -6, cmd '/opt/ros/rolling/lib/rclcpp_components/component_container_isolated --ros-args --log-level info --ros-args -r __node:=nav2_container --params-file /tmp/launch_params_xwa4_atv --params-file /tmp/launch_params_3tgew0_b -r /tf:=tf -r /tf_static:=tf_static'].
```

Yadunund commented:
It seems like https://github.com/ros2/rmw_zenoh/pull/293/commits/082e64c42e5ee09c0f17b667001e306b4ff42e20 introduced this behavior as I can't reproduce the crash if I drop that commit.

Yadunund committed:
Fix -Wnon-pod-varargs

Signed-off-by: Yadunund <yadunund@gmail.com>

ahcorde reviewed (changes requested):
conflicts

clalancette committed:
Make PublisherData::copy_gid have the same signatures as others.

This just makes it more consistent with the rest of the
APIs.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
> conflicts

I've now rebased it onto the latest.

clalancette commented:
Besides rebasing this, I also did a local test where I rebased this onto https://github.com/ros2/rmw_zenoh/pull/307 , and also reverted 19b84e98771b0ec07c16959e1d746c51fdab4fc6 .  But I am running into shutdown issues in `client_data_handler` where it is deadlocking, so I'll need to look into that further.

Yadunund committed:
Add api to init client

Signed-off-by: Yadunund <yadunund@gmail.com>

clalancette committed:
Vastly simplify handling of in-flight ClientData.

We basically use the pointer as a key to lookup the
shared_ptr, and then always use the shared_ptr.  That
ensures that even if the pointer was dropped from the
node while the client callback is operating, it will
still be valid and won't be destroyed until that method
returns.

Signed-off-by: Chris Lalancette <clalancette@gmail.com>

clalancette commented:
All right, I've just rebased this and also pushed another commit to vastly simplify our handling of in-flight data.  I think this should work a lot more reliably now.  With that, I'm going to approve and merge this.  Thanks for the long slog on this one.
